### PR TITLE
Notify on schedule updates and deletions

### LINKED ISSUE #2613: Add an option to notify subscribers when scheduled maintenance date changes
is there an option to notify subscribers when scheduled maintenance date changes. if not can this be added.

Cachet v2.3.11

### PR DESCRIPTION
**Pull request by atymic**
WIP.

I'm new to this project, so I assume there will be some things i've missed or style differences.

I also have to look into adding tests :)

Closes #2613

### CODE REVIEW

**Comment by jbrooksuk** [comment]
Thanks for this!

I'd probably rename `SendScheduleEmailNotificationHandler.php` and have three handlers for the different events. This means we can remove the if conditions around the type. Think SOLID :)

**Comment by atymic** [comment]
> Thanks for this!
> 
> I'd probably rename `SendScheduleEmailNotificationHandler.php` and have three handlers for the different events. This means we can remove the if conditions around the type. Think SOLID :)

No worries, i'll update it soon 😄

File: resources/views/dashboard/maintenance/edit.blade.php [hunk 1]
```diff
@@ -56,6 +56,15 @@
                         <label>{{ trans('forms.schedules.completed_at') }}</label>
                         <input type="text" name="completed_at" class="form-control flatpickr-time" data-date-format="Y-m-d H:i" value="{{ $schedule->completed_at_datetimepicker }}" placeholder="{{ trans('forms.schedules.completed_at') }}">
                     </div>
+                    @if($notificationsEnabled)
+                    <input type="hidden" name="notify" value="0">
```
[the remark this answers is not in the document]
Comment (atymic) [reply resources/views/dashboard/maintenance/edit.blade.php]: There is a label here :)

**Comment by atymic** [comment]
I updated the dashboard to add a `notify` option on the edit, as well as updating the commands.
Any suggestions on what do do on the delete? Currently there's just a sweetalert confirm, but that means we can't let them pick a notify option.

**Comment by jbrooksuk** [comment]
Can we add three buttons to Sweetalert? I can't remember!

**Comment by atymic** [comment]
Currently we are using a helper to show the dialog:
```blade
 <a href="{{ cachet_route('dashboard.schedule.delete', [$incident->id], 'delete') }}" class="btn btn-danger confirm-action" data-method='DELETE'>{{ trans('forms.delete') }}</a>
```

The code that is called to show the dialog is pretty focused on just confirm dialogs.

Should I add some custom JS to create a sweet alert with three buttons (Cancel, Delete, Delete & Notify)? Or this there a different way you would prefer?

```javascript
function askConfirmation(callback, cancelCallback) {
        swal({
            type: "warning",
            title: "Confirm your action",
            text: "Are you sure you want to do this?",
            buttonsStyling: false,
            reverseButtons: true,
            confirmButtonText: "Yes",
            confirmButtonClass: "btn btn-lg btn-danger",
            cancelButtonClass: "btn btn-lg btn-default",
            showCancelButton: true,
            focusCancel: true
        }).then(function () {
            if (_.isFunction(callback)) callback();
        }, function () {
            if (_.isFunction(cancelCallback)) cancelCallback();
        });
    }
```

**Comment by jbrooksuk** [comment]
You could create a custom helper, that's fine :)

**Comment by atymic** [comment]
Sorry for the delay @jbrooksuk.

I've pushed up the code to add a sweet alert dialog with a checkbox for selecting notification preference.

What is the procedure for making JS changes? Should I commit all of the mix output file or does the CI build them?

File: resources/views/dashboard/maintenance/index.blade.php [hunk 2]
```diff
@@ -31,7 +31,14 @@
                         </div>
                         <div class="col-xs-6 text-right">
                             <a href="{{ cachet_route('dashboard.schedule.edit', [$incident->id]) }}" class="btn btn-default">{{ trans('forms.edit') }}</a>
-                            <a href="{{ cachet_route('dashboard.schedule.delete', [$incident->id], 'delete') }}" class="btn btn-danger confirm-action" data-method='DELETE'>{{ trans('forms.delete') }}</a>
+                            <a href="#" class="btn btn-danger delete-schedule">
+                                {{ trans('forms.delete') }}
+                                <form action='{{ cachet_route('dashboard.schedule.delete', [$incident->id], 'delete') }}' method='POST' style='display:none'>
+                                    <input type='hidden' name='_method' value='DELETE'>
```
[the remark this answers is not in the document]
Comment (atymic) [reply resources/views/dashboard/maintenance/index.blade.php]: Not a user input

File: resources/views/dashboard/maintenance/index.blade.php [hunk 3]
```diff
@@ -31,7 +31,14 @@
...
+                                    <input type='hidden' name='_token' value='{{ csrf_token() }}'>
```
[the remark this answers is not in the document]
Comment (atymic) [reply resources/views/dashboard/maintenance/index.blade.php]: Not a user input

File: resources/views/dashboard/maintenance/index.blade.php [hunk 4]
```diff
@@ -31,7 +31,14 @@
...
+                                    <input type="hidden" name="notify" value="0">
```
[the remark this answers is not in the document]
Comment (atymic) [reply resources/views/dashboard/maintenance/index.blade.php]: Not a user input

**Comment by atymic** [comment]
Also, build seems to be failing due to SSL errors on the security check? 🤷‍♂ 

```
In ErrorChunk.php line 105:
                                                                    
  SSL connect error for "https://security.symfony.com/check_lock".  
                                                                    
```

**Comment by jbrooksuk** [comment]
We'll build assets, don't worry :)

**Comment by pmkakaci** [comment]
Hello guys,
Is this request still on work ? I'm pretty interested about this one.
Kind regards,

**Comment by Florian-94** [comment]
Hello,
I tried to add/replace all of these 25 files in my docker cachet image (tag 2.4.0-dev   sha 1b1bcb00e588c6b43249a3a875140ecedca189c5a38f1b72593eb7415c1efde6).
But I still doesn't receive mail when I create a new maintenance. It works for incident but not maintenance. It's a pity !
Do you know how to fix it ? (or do you know when this bug will be fix in a new docker image ?).
Thanks !

**Pull request closed without merging**